d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,390 @@
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.d2.discovery.util.LogUtil.*;
+
+
+/**
+ * Update the state of the RelativeLoadBalancerStrategy periodically
+ */
+public class RelativeStateUpdater implements StateUpdater
+{
+  private static final Logger LOG = LoggerFactory.getLogger(RelativeStateUpdater.class);
+  public static final double MIN_HEALTH_SCORE = 0.0;
+  public static final double MAX_HEALTH_SCORE = 1.0;
+  private static final double SLOW_START_INITIAL_HEALTH_SCORE = 0.01;
+  private static final int SLOW_START_RECOVERY_FACTOR = 2;
+  private static final int LOG_UNHEALTHY_CLIENT_NUMBERS = 10;
+  private static final long EXECUTOR_INITIAL_DELAY = 10;
+
+  private final D2RelativeStrategyProperties _relativeStrategyProperties;
+  private final QuarantineManager _quarantineManager;
+  private final ScheduledExecutorService _executorService;
+  private final List<PartitionLoadBalancerStateListener.Factory<PartitionRelativeLoadBalancerState>> _listenerFactories;
+
+  private ConcurrentMap<Integer, PartitionRelativeLoadBalancerState> _partitionLoadBalancerStateMap;
+
+  RelativeStateUpdater(D2RelativeStrategyProperties relativeStrategyProperties,
+                       QuarantineManager quarantineManager,
+                       ScheduledExecutorService executorService,
+                       List<PartitionLoadBalancerStateListener.Factory<PartitionRelativeLoadBalancerState>> listenerFactories)
+  {
+    _relativeStrategyProperties = relativeStrategyProperties;
+    _quarantineManager = quarantineManager;
+    _executorService = executorService;
+    _listenerFactories = listenerFactories;
+    _partitionLoadBalancerStateMap = new ConcurrentHashMap<>();
+
+    _executorService.scheduleWithFixedDelay((Runnable) this::updateState, EXECUTOR_INITIAL_DELAY,
+        _relativeStrategyProperties.getUpdateIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void updateState(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    if (!_partitionLoadBalancerStateMap.containsKey(partitionId))
+    {
+      // If the partition is not initialized, initialize the state synchronously
+      initializePartition(trackerClients, partitionId, clusterGenerationId);
+    } else if (clusterGenerationId != _partitionLoadBalancerStateMap.get(partitionId).getClusterGenerationId())
+    {
+      // Asynchronously update the state if it is from uri properties change
+      PartitionRelativeLoadBalancerState oldPartitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      _executorService.execute(() -> updateStateForPartition(trackerClients, partitionId, oldPartitionState, clusterGenerationId));
+    }
+  }
+
+  @Override
+  public Ring<URI> getRing(int partitionId)
+  {
+    return _partitionLoadBalancerStateMap.get(partitionId) == null
+        ? null
+        : _partitionLoadBalancerStateMap.get(partitionId).getRing();
+  }
+
+  Map<URI, Integer> getPointsMap(int partitionId)
```

Comment:
consider adding this to API so you don't need to cast to call this.

Reply:
This is only used by tests, the actual implementation only needs to access the ring. That's why I didn't put it to the interface